src/app/repositories/firebase/schema_firebase_repository.py

```diff
@@ -28,9 +32,43 @@ def get_latest_schema_with_survey_id(
             .stream()
         )
 
-    def create_schema(
-        self, schema_id: str, schema_metadata: SchemaMetadataWithoutGuid
-    ) -> SchemaMetadata:
+    def perform_new_schema_transaction(
+        self,
+        schema_id: str,
+        next_version_schema_metadata: SchemaMetadata,
+        schema: Schema,
+        stored_schema_filename: str,
+    ) -> None:
+        """
+        A transactional function that wrap schema creation and schema storage processes
+
+        Parameters:
+        schema_id (str): The unique id of the new schema.
+        next_version_schema_metadata (SchemaMetadata): The schema metadata being added to firestore.
+        schema (Schema): The schema being stored.
+        stored_schema_filename (str): Filename of uploaded json schema.
+        """
+
+        # A stipulation of the @firestore.transactional decorator is the first parameter HAS
+        # to be 'transaction', but since we're using classes the first parameter is always
+        # 'self'. Encapsulating the transaction within this function circumvents the issue.
+        @firestore.transactional
+        def post_schema_transaction_run(transaction: Transaction):
+            self.create_schema_in_transaction(
```

Comment:
Do we need to call a separate private function here? I think it would be just as clear to call `transaction.set()` directly inside of this function since it's just a one-liner and it's pretty clear what operation is being performed

Reply:
I can do that. But my though is I treat this as a function wrapper that only calls other functions into it, so it looks more consistent in the create schema in db process and store schema process

Reply:
I can see that, but I think doing it this way makes it clearer that `create_schema_in_transaction` is actually a transaction, as in it's performing operations on the database. Storing schema in the bucket isn't a transaction, so it makes sense to wrap it in a function call. If you wrap both in a function at first glance the storing schema function might seem like it's a transaction operation.

If you still feel like this is the best way then I'm happy to approve it, but personally I think it's an unnecessary extra function to maintain

Reply:
I see your point. My view is that this arrangement can be easier to maintain in terms of scalability that I can reuse the functions in other transactions or add other transactional functions into the wrapper. Not so obvious in the current situation but probably for possible future. I know it is a bit of overthinking here, but kind of like to have a structure that allows that to happen.

Reply:
Fair enough, although it's always good to bear in mind YAGNI (you ain't gonna need it) when you're trying to future proof something. It might be better to introduce a structure here when the new feature comes in that requires it, rather than trying to do it in advance and have extra code to maintain